scheduling/storage_providers/etcd.go

```diff
@@ -39,20 +69,206 @@ func (etcd *ETCDStorageProvider) Get(key string, prefix bool) ([]string, error)
 	if err != nil {
 		return nil, fmt.Errorf("failed to get keys with prefix %s: %w", key, err)
 	}
 
-	var result []string
 	for _, kv := range resp.Kvs {
-		result = append(result, string(kv.Value))
+		result[string(kv.Key)] = string(kv.Value)
 	}
 	if len(result) == 0 {
 		return nil, &KeyNotFoundError{Key: key}
 	}
 	return result, nil
 }
+
+func (etcd *ETCDStorageProvider) Delete(key string, lock LockObject) error {
+	if key == "" {
+		return fmt.Errorf("key is empty")
+	}
+
+	// get lock and check if it is valid
+	lockKeyPath := fmt.Sprintf("/%s/%s", LOCKPREFIX, strings.TrimLeft(key, "/"))
+	resp, err := etcd.client.Get(etcd.ctx, lockKeyPath)
+	if err != nil {
+		return fmt.Errorf("failed to get key %s: %w", lockKeyPath, err)
+	}
+
+	if len(resp.Kvs) == 0 {
+		return fmt.Errorf("key is not locked")
+	}
+
+	lockInfo := LockInformation{}
+	if err := lockInfo.Unmarshal(resp.Kvs[0].Value); err != nil {
+		return fmt.Errorf("failed to unmarshal lock information: %w", err)
+	}
+	if lockInfo.ID != lock.ID {
+		return fmt.Errorf("key is locked by another id: locked by: %s, unlock  by: %s", lockInfo.ID, lock.ID)
+	}
+
+	_, err = etcd.client.Delete(etcd.ctx, key)
+	if err != nil {
+		return fmt.Errorf("failed to delete key %s: %w", key, err)
+	}
+
+	_, err = etcd.client.Delete(etcd.ctx, lockKeyPath)
+	if err != nil {
+		return fmt.Errorf("failed to delete key %s: %w", lockKeyPath, err)
+	}
+
+	// close the lock channel
+	if lock.Channel != nil {
+		close(*lock.Channel)
+	}
+	return nil
+}
+
+func (etcd *ETCDStorageProvider) ListKeys(prefix string) ([]string, error) {
+	resp, err := etcd.client.Get(etcd.ctx, prefix, clientv3.WithPrefix())
+	if err != nil {
+		return nil, fmt.Errorf("failed to get keys with prefix %s: %w", prefix, err)
+	}
+
+	var result []string
+	for _, kv := range resp.Kvs {
+		result = append(result, string(kv.Key))
+	}
+	sort.Strings(result)
+
+	return result, nil
+}
+
+func (etcd *ETCDStorageProvider) Lock(key string) (LockObject, error) {
+	if key == "" {
+		return LockObject{}, fmt.Errorf("key is empty")
+	}
+	lockKeyPath := fmt.Sprintf("/%s/%s", LOCKPREFIX, strings.TrimLeft(key, "/"))
+
+	// check if the key is already locked
+	resp, err := etcd.client.Get(etcd.ctx, lockKeyPath)
+	if err != nil {
+		return LockObject{}, fmt.Errorf("failed to get key %s: %w", lockKeyPath, err)
+	}
+
+	if len(resp.Kvs) != 0 {
+		lock := LockInformation{}
+		if err := lock.Unmarshal(resp.Kvs[0].Value); err != nil {
+			return LockObject{}, fmt.Errorf("failed to unmarshal lock information: %w", err)
+		}
+		if time.Since(lock.Date) < ValidTimePeriod {
+			return LockObject{}, fmt.Errorf("key is already locked by: %s", lock.ID)
+		}
+	}
+
+	id := uuid.New().String()
+	lock := LockInformation{
+		ID:   id,
+		Key:  key,
+		Date: time.Now().UTC(),
+	}
+	data, err := lock.Marshal()
+	if err != nil {
+		return LockObject{}, fmt.Errorf("failed to marshal lock information: %w", err)
+	}
+
+	// unmarshal the lock information
+	newLock := LockInformation{}
```

Comment:
why are we marshalling then unmarshalling?

Reply:
did it for debugging purposes, need to remove it.